### DISCUSSION TITLE
[Question]: Developing on docker with live reload

### DISCUSSION THREAD

**Discussion by noobskies** [danny-avila/LibreChat#475] [Q&A] [answered]
### Contact Details

t.noobskie@gmail.com

### What is your question?

I'm currently trying to add some new features however I am not entirely sure how the dev workflow works, are you using docker to start the app on dev?

### More Details

Main issue is I just want to enable live reload so when i make ui changes it refreshes, i'm not sure if you already have this implemented or not

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
Awesome, thanks for wanting to contribute! I recommend jumping in our discord as the active devs chat there a lot: https://discord.gg/CEe6vDg9Ky, ping me when you join `@dannya`

To answer your question, you shouldn't be using docker for dev work. I'm sure there's an easy way to configure it, but docker has not been configured for development.

The best way to is to install locally, which will depend on your environment but in general it goes like this, provided you have env file setup, the latest npm installed and mongodb installed or using atlas; note: you can use mongodb from your docker container on your local setup, I can get into it more if you want.
```bash
# in the directory you want the project in
git clone https://github.com/danny-avila/LibreChat.git
cd ./LibreChat
npm ci
npm run backend #backend-dev if you are making changes to the api
npm run frontend-dev # this is for live reload
```

**Reply by noobskies** [reply]
not sure why this is happening any ideas?

```
There was an uncaught error: EACCES: permission denied, open './data/cache.json'
Clearing sync timeouts before exiting...
[nodemon] app crashed - waiting for file changes before starting...
```

I'm using `npm run backend-dev` everything works but i get that error after the first message finishes

**Reply by ClaraLeigh** [reply]
File permissions perhaps? This can happen when working with docker/host in parallel, just make sure you own the file and that you have read/write perms

**Reply by thunderbug1** [reply]
> Awesome, thanks for wanting to contribute! I recommend jumping in our discord as the active devs chat there a lot: https://discord.gg/CEe6vDg9Ky, ping me when you join `@dannya`
> 
> To answer your question, you shouldn't be using docker for dev work. I'm sure there's an easy way to configure it, but docker has not been configured for development.
> 
> The best way to is to install locally, which will depend on your environment but in general it goes like this, provided you have env file setup, the latest npm installed and mongodb installed or using atlas; note: you can use mongodb from your docker container on your local setup, I can get into it more if you want.
> 
> ```shell
> # in the directory you want the project in
> git clone https://github.com/danny-avila/LibreChat.git
> cd ./LibreChat
> npm ci
> npm run backend #backend-dev if you are making changes to the api
> npm run frontend-dev # this is for live reload
> ```

I love the devcontainer feature of VS code. It makes it really easy to create a reproducible development environment for all developers.
Normally it is quite simple to set it up and fully develop in a docker container. In this project we might need a docker compose file since there are multiple services involved but that is also supported as far as I see: https://github.com/microsoft/vscode-dev-containers/blob/main/container-templates/docker-compose/.devcontainer/docker-compose.yml

**Reply by thunderbug1** [reply, marked as answer]
I added a devcontainer setup, see here: https://github.com/danny-avila/LibreChat/pull/524
